### ISSUE TITLE
Port to JavaScript

### ISSUE THREAD

**Issue by jyn514** [charlesdaniels/bitshuffle#58]
To avoid heavy-duty hosting like Heroku, it would be nice to have this in native JavaScript. This will also have the benefit of allowing us to host on Github Pages, which only supports static sites.

Waiting on ~~https://github.com/charlesdaniels/bitshuffle/issues/57 and https://github.com/charlesdaniels/bitshuffle/issues/54~~ https://github.com/charlesdaniels/bitshuffle/issues/51.

**Comment by charlesdaniels** [comment]
It will also be a good way to ensure our spec is really airtight once I get around to writing it. 

We should investigate python to js transpilers before re writing everything from scratch though. 

- - - - 

This email was sent from my smartphone.

**Comment by z2oh** [comment]
I'd be happy to tackle this once the spec is written.

Options are write it to Rust and use WASM (which is cooler and will result in a much faster execution time) or write it in JS. Assuming adequate library support, this shouldn't take too long to rewrite in JS. In fact, I imagine less time would be spent rewriting it than configuring a transplier to work (though transpilation has the added benefit of generating the js code as part of a build step, meaning only one code implementation needs to be maintained.)

However, manually porting the implementation to another language should provide a good test on the robustness of the spec.

**Comment by jyn514** [comment]
I would love to see this in Rust. What would a WASM compiler look like? The advantage would be now that it's in both JS, Rust, and Python :D

**Comment by charlesdaniels** [comment]
I am in favor of Rust as well. The only libraries we should need would be base64, bz2, gzip, and any other compression algorithms we support... I don't think the list is very long, and all of these have C implementations we can link against even if Rust does not have native ports.

**Comment by jyn514** [comment]
(and sha256)
